src/PhpDoc/TypeNodeResolver.php

```diff
@@ -379,32 +389,46 @@ private function resolveGenericTypeNode(GenericTypeNode $typeNode, NameScope $na
 					}
 				}
 
-				if (!$mainType->isIterable()->yes()) {
-					return new GenericObjectType($mainType->getClassName(), $genericTypes);
-				}
-
-				if (
-					count($genericTypes) !== 1
-					|| $classReflection->getTemplateTypeMap()->count() === 1
-				) {
-					return new GenericObjectType($mainType->getClassName(), $genericTypes);
-				}
+//				if (!$mainType->isIterable()->yes()) {
+//					return new GenericObjectType($mainType->getClassName(), $genericTypes);
+//				}
+//
+//				if (
+//					count($genericTypes) !== 1
+//					|| $classReflection->getTemplateTypeMap()->count() === 1
+//				) {
+//					return new GenericObjectType($mainType->getClassName(), $genericTypes);
+//				}
 			}
 		}
 
-		if ($mainType->isIterable()->yes()) {
-			if (count($genericTypes) === 1) { // Foo<ValueType>
-				return TypeCombinator::intersect(
-					$mainType,
-					new IterableType(new MixedType(true), $genericTypes[0])
-				);
-			}
-
-			if (count($genericTypes) === 2) { // Foo<KeyType, ValueType>
-				return TypeCombinator::intersect(
-					$mainType,
-					new IterableType($genericTypes[0], $genericTypes[1])
-				);
+//		if ($mainType->isIterable()->yes()) {
+//			if (count($genericTypes) === 1) { // Foo<ValueType>
+//				return TypeCombinator::intersect(
+//					$mainType,
+//					new IterableType(new MixedType(true), $genericTypes[0])
+//				);
+//			}
+//
+//			if (count($genericTypes) === 2) { // Foo<KeyType, ValueType>
+//				return TypeCombinator::intersect(
+//					$mainType,
+//					new IterableType($genericTypes[0], $genericTypes[1])
+//				);
+//			}
+//		}
+
+		$templateTypeMap = array_values($classReflection->getTemplateTypeMap()->getTypes());
+		foreach ($genericTypes as $i => &$genericType) {
+			if ($genericType instanceof ConstantStringType && $genericType->getValue() === 'star' && array_key_exists($i, $templateTypeMap)) {
+					$templateType = $templateTypeMap[$i];
+					assert($templateType instanceof TemplateType);
+					$genericType = TemplateTypeFactory::create(
```

Comment:
Here we would ideally stread TemplateStarType with this template type as an argument. Basically all functions would be a proxy to this type.